man/dmuuid.1
.TH "DMUUID" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmuuid \(en generates pseudo\-random UUIDs
.SH SYNOPSIS
\f[B]dmuuid\f[R] \-\-help
.PP
\f[B]dmuuid\f[R] \-\-version
.PP
\f[B]dmuuid\f[R] [\-\-\f[B]count\f[R] \f[I]n\f[R]]
[\-\-\f[B]hyphens\f[R]]
.PP
\f[B]dmuuid\f[R] \-\-\f[B]convert\f[R] < \f[I]file\f[R]
.SH DESCRIPTION
The \f[B]dmuuid\f[R] program generates pseudo\-random UUIDs.
By default, DMPACK uses 32 characters long UUIDv4 identifers in
hexadecimal format (without hyphens).
If command\-line flag \f[CR]\-\-hyphens\f[R] is set, 36 characters long
identifiers with hyphens are returned.
The option \f[CR]\-\-convert\f[R] expects UUIDs to be passed via
standard input.
Invalid identifiers will be replaced with the default UUIDv4.
.SH OPTIONS
.TP
\f[B]\-\-convert\f[R], \f[B]\-c\f[R]
Add hyphens to hexadecimal UUIDs passed via \f[I]stdin\f[R].
.TP
\f[B]\-\-count\f[R], \f[B]\-n\f[R] \f[I]n\f[R]
Number of UUIDs to generate (default: 1).
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-hyphens\f[R], \f[B]\-p\f[R]
Return 36 characters long UUIDs with hyphens.
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
UUID generation was successful.
.TP
\f[B]1\f[R]
Failure.
UUID generation failed.
.SH EXAMPLE
Create three random identifiers:
.IP
.EX
$ dmuuid \-n 3
6827049760c545ad80d4082cc50203e8
ad488d0b8edd4c6c94582e702a810ada
3d3eee7ae1fb4259b5df72f854aaa369
.EE
.PP
Add hyphens to a hexadecimal UUIDv4:
.IP
.EX
$ echo \(dq3d3eee7ae1fb4259b5df72f854aaa369\(dq | dmuuid \-c
3d3eee7a\-e1fb\-4259\-b5df\-72f854aaa369
.EE


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dmuuid.f90
! dmuuid.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmuuid
    !! Generates pseudo-random UUIDs. By default, DMPACK uses UUIDv4 in
    !! hexadecimal format, i.e., without hyphens. By using the command-line
    !! argument `--hyphens`, these are added.
    !!
    !! The command-line flag `--convert` expects UUIDs to be passed via
    !! standard input. Invalid UUIDs will be replaced with the default UUIDv4.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmuuid'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    type :: app_type
        !! Command-line arguments.
        integer :: count   = 1       !! Number of UUIDs to generate (>= 1).
        logical :: convert = .false. !! Add hyphens to 32 characters long UUIDv4.
        logical :: hyphens = .false. !! Add hyphens to generated UUIDs.
    end type app_type

    integer        :: i, rc
    type(app_type) :: app

    ! Initialise DMPACK.
    call dm_init()

    ! Get command-line arguments.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Read UUIDv4s from stdin and add hyphens.
    if (app%convert) then
        call convert()
        call dm_stop(STOP_SUCCESS)
    end if

    ! Generate and output UUIDs.
    do i = 1, app%count
        if (app%hyphens) then
            print '(a36)', dm_uuid4_hyphens()
        else
            print '(a32)', dm_uuid4()
        end if
    end do
contains
    subroutine convert()
        !! Reads UUIDs from standard input, adds hyphens, and prints the result
        !! to standard output.
        character(UUID_LEN) :: uuid
        integer             :: stat

        do
            read (*, '(a)', iostat=stat) uuid
            if (is_iostat_end(rc)) exit
            if (stat /= 0 .or. .not. dm_uuid4_is_valid(uuid)) uuid = UUID_DEFAULT
            print '(a)', dm_uuid4_hyphenize(uuid)
        end do
    end subroutine convert

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments.
        type(app_type), intent(out) :: app !! App type

        type(arg_parser_class) :: parser

        call parser%add('convert', short='c', type=ARG_TYPE_LOGICAL) ! -c, --convert
        call parser%add('count',   short='n', type=ARG_TYPE_INTEGER) ! -n, --count
        call parser%add('hyphens', short='p', type=ARG_TYPE_LOGICAL) ! -p, --hyphens

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('convert', app%convert)
        call parser%get('count',   app%count)
        call parser%get('hyphens', app%hyphens)

        rc = validate(app)
    end function read_args

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(inout) :: app !! App type.

        rc = E_INVALID

        if (app%convert .and. app%hyphens) then
            call dm_error_out(rc, 'argument --convert conflicts with --hyphens')
            return
        end if

        if (app%convert .and. app%count /= 1) then
            call dm_error_out(rc, 'argument --convert conflicts with --count')
            return
        end if

        if (app%count <= 0) then
            call dm_error_out(rc, 'invalid count')
            return
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
    end subroutine version_callback
end program dmuuid
